### PR TITLE
Add subcommand for collecting files by a ruleset of patterns

### PR DESCRIPTION
**Pull request by johanneskoester**
This is useful for generating sample and unit sheets. E.g., you can have this input sheet (here `unit-patterns.tsv`):
```
S743_Nr(?P<nr>[0-9]+)  S743_1/01_fastq/S743Nr{nr}.*.fastq.gz
S839_Nr(?P<nr>[0-9]+)   S839_*/01_fastq/S839Nr{nr}.*.fastq.gz
S888_Nr(?P<nr>[0-9]+)   S888/S888_1/01_fastq/S888Nr{nr}.*.fastq.gz
```

And run 
```
cut -f1 samples.tsv | tail -n+2 | snakedeploy collect-files --config unit-patterns.tsv
```
with samples.tsv being a sample sheet where we cut out the first columns containing the sample ids. The config sheet above lets us select, for each sample, a glob pattern, which is then used to obtain the files on disk that correspond to this sample, which are then printed tab separated to STDOUT, along with the sample id that we put in.

### CODE REVIEW

File: snakedeploy/collect_files.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+            filter(
+                lambda match: match.match is not None, get_matches(item, config_sheet)
+            )
+        )
+
+        if not matches:
+            raise UserError(f"No input pattern in config sheet matches {item}.")
+        elif len(matches) > 1:
+            raise UserError(f"Item {item} matches multiple input patterns.")
+        else:
```
Comment (vsoch) [inline snakedeploy/collect_files.py:29]: For this third else, we can probably remove it and move the rest of the function one indentation out. If either of the first two blocks are triggered, raising the error will stop execution (so I don't think we would need the else?) E.g., in this small example we never get to the print statement:

```python
from snakedeploy.exceptions import UserError

if 2==1:
   print("Case 1")

# This one will trigger
elif 1==1:
    raise UserError('blah')

# We never get here
print('NO')
```

File: snakedeploy/exceptions.py [hunk 2]
```diff
@@ -31,3 +31,7 @@ class UnrecognizedProviderError(ValueError):
 
     def __init__(self, message, *args, **kwargs):
         super(UnrecognizedProviderError, self).__init__(*args, **kwargs)
+
+
+class UserError(Exception):
```
Comment (vsoch) [inline snakedeploy/exceptions.py:36]: Is the base Exception class the right one to use here (e.g., and not ValueError if it's typically something that the user did?"

File: snakedeploy/version.py [hunk 3]
```diff
@@ -15,7 +15,10 @@
 # Global requirements
 
 
-INSTALL_REQUIRES = (("requests", {"min_version": None}),)
+INSTALL_REQUIRES = (
+    ("requests", {"min_version": None}),
+    ("pandas", {"min_version": None}),
```
Comment (vsoch) [inline snakedeploy/version.py:20]: pandas is a fairly big dependency to add - and known for fairly large changes between it's versions. If you intend to have many other uses for it around the library then we could keep, but if it's just needed to read a csv and shuffle some values around, could the same not be done with the `csv` library (and avoid the heavy dependency?)

**Review by vsoch — COMMENTED** [review]
@johanneskoester this is a neat addition! I have some suggestions (that I can offer to help with too) and am interested in your thoughts:

1. First, see my comments - there is one section that I think can be un-indented, etc.
2. I'm wondering if we can refactor to not add pandas as a dependency, as (so far) it's a fairly small functionality that could be done with `csv`. Pandas is quite a large hairy dependency that I think should be added only if absolutely necessary.
3. We should have some docs for this - can you point me to a samples.tsv that would work with your example in the description above and I'll add it to the docs in this PR?
4. It looks like the formatting issue is just due to black versioning, which should be `black==20.8b1` If I can help on any of the above, I can definitely run this version locally.
5. It would be good to add a test that this works as expected (I can offer to do this too with samples.tsv).

File: snakedeploy/exceptions.py (hunk 2 above)
Comment (johanneskoester) [reply snakedeploy/exceptions.py:36]: I think the base exception should be ok, but if you prefer a ValueError that would be fine for me as well. Your choice.

Comment (vsoch) [reply snakedeploy/exceptions.py:36]: okay let's do ValueError.

Comment (vsoch) [inline snakedeploy/exceptions.py:36]:
```suggestion
class UserError(ValueError):
```

**Pull request merged**